[FEATURE] Table legend for time series chart panel

juliepagano opened the pull request:
The time series chart panel now supports the option to select a legend "mode" that is either a list (previous legend UI) or a table (new option).

<!--
  See the contributing guide for detailed guidance about contributing.
  https://github.com/perses/perses/blob/main/CONTRIBUTING.md
-->

# Notes for reviewers

- I added a dashboard that exercises the various modes at http://localhost:3000/projects/testing/dashboards/TimeSeriesLegends?start=5m if helpful for testing these changes. There's also a bunch in storybook.
- I'm intentionally leaving "list" mode alone at this time to be minimally disruptive until we add more sizes and other functionality.

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [x] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [x] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [x] Visual tests are stable and unlikely to be flaky. See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests) and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

# Screenshots

<img width="1521" alt="image" src="https://github.com/perses/perses/assets/396962/561108d8-fd6c-40f5-b763-bc2fc052e27d">
<img width="1521" alt="image" src="https://github.com/perses/perses/assets/396962/15ecb530-a246-420a-92c9-16602cace0e9">


https://github.com/perses/perses/assets/396962/d65628ee-7945-4db3-bdc5-894d15640def

juliepagano commented on ui/components/src/ContentWithLegend/model/content-with-legend-model.test.ts:
```diff
@@ -36,118 +37,124 @@ describe('getContentWithLegendLayout', () => {
     });
   });
 
```
I put a loop around these tests to check both list and table, which made the diff a little funky, but the tests haven't changed much since the previous PR.

juliepagano commented on ui/components/src/ContentWithLegend/model/content-with-legend-model.ts:
```diff
@@ -112,16 +127,26 @@ export function getContentWithLegendLayout({
+    }
+  } else {
+    legendWidth = position === 'Right' ? TABLE_LEGEND_SIZE['Right'] : width;
+    legendHeight = position === 'Bottom' ? TABLE_LEGEND_SIZE['Bottom'] * 26 : height;
```
TODO: get the row height from the legend utils instead of hardcoding it here.

juliepagano commented on ui/components/src/ContentWithLegend/model/content-with-legend-model.ts:
```diff
@@ -112,16 +127,26 @@ export function getContentWithLegendLayout({
+  let legendHeight;
+
+  if (mode === 'List') {
+    // TODO: normalize list to share similar height options as the table
+    // when we add more size options.
```
This avoids this being really disruptive to current list users who were used to the previous size. We default the table to what will be "medium" when we add the new sizes.

juliepagano commented on ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanelComponent.stories.tsx:
```diff
@@ -89,19 +100,10 @@ const meta: Meta<typeof TimeSeriesChart.PanelComponent> = {
-                      value: '1',
-                    },
-                  ],
+                body: mockTimeSeriesResponseWithManySeries({
```
Changed this to a mock that shows a bunch of series, which is better for showing off the legends.

juliepagano commented on schemas/panels/time-series/mig.cuepart:
```diff
@@ -7,12 +7,20 @@ if #panel.type == "timeseries" || #panel.type == "graph" {
             	  	if #panel.options.legend.placement == "bottom" { "Bottom" },
             	  	if #panel.options.legend.placement == "right" { "Right" },
             	  ][0]
```
This is my first time changing these mig.cuepart files, and I wasn't 100% sure how to test that they work. Any pointers?

juliepagano replied:
Aha, I think I found some failing go tests that answer my question here.

juliepagano commented on schemas/panels/time-series/time-series.cue:
```diff
@@ -19,6 +19,7 @@ import (
 
 #legend: {
 	position: "Bottom" | "Right"
+	mode?: "List" | "Table"
```
Making this optional, so the addition of the table mode isn't a breaking change for consumers.

juliepagano commented on internal/api/shared/migrate/testdata/simple_grafana_dashboard.json:
```diff
@@ -332,7 +332,7 @@
             "legend": {
               "calcs": [],
-              "displayMode": "list",
+              "displayMode": "table",
```
Swapping one of these to "table" so we exercise both list and table modes.

juliepagano commented on ui/components/src/ContentWithLegend/ContentWithLegend.stories.tsx:
```diff
@@ -140,25 +142,37 @@ export const Position: Story = {
-                }}
-              />
+              <Stack spacing={1} direction="row" flexWrap="wrap">
+                {legendPositions.map((position) => {
```
Added positions so you can see the combo of all modes and all positions.

juliepagano commented on ui/components/src/ContentWithLegend/model/content-with-legend-model.ts:
```diff
@@ -113,15 +128,30 @@ export function getContentWithLegendLayout({
+    // when we add more size options.
+    legendWidth = position === 'Right' ? 200 : width;
+
+    // TODO: account for number of legend items returned when adjusting legend spacing
```
Logic for the "list" mode is largely unchanged besides the vars being declared above.

juliepagano commented on dev/data/dashboard.json:
```diff
@@ -3057,5 +3057,192 @@
       ]
     }
+  },
+  {
```
I didn't add any e2e tests for this because we get pretty good coverage from jest tests and visual tests from storybook. However, I added a test dashboard for ease of debugging and playing around with legends in the app context.

eunicorn commented:
Is this expected? Looks a bit weird with the graph being so squished

![image](https://github.com/perses/perses/assets/9817826/64dbdb44-1930-42be-b756-bfd89090900e)

juliepagano commented:
> Is this expected? Looks a bit weird with the graph being so squished
> 
> ![image](https://user-images.githubusercontent.com/9817826/242734728-64dbdb44-1930-42be-b756-bfd89090900e.png)

It is currently. The expectation is that at the "medium" size (what we're starting with), the bottom-positioned table is 5 rows tall. We'll handle adding in a "small" size in a follow-up. As part of that, we should also probably add some special handling to make it smaller if there are less than 5 rows (there's been a TODO comment to do something similar for the list legend for ages). The right-positioned table looks a little nicer at this point.

There's also some open questions about how big the panel rendering should be in the settings, so it can better reflect what you might see when you put it onto the dashboard. Some conversations about that prompted @cndonovan to start on https://github.com/perses/perses/pull/1208. 

TLDR: we should do more here in follow-ups, but I'm hoping this PR can be incremental progress for delivering the table legend. Let me know if you think it's critical to resolve some of these issues in this PR instead of in follow-ups.

sjcobb commented:
The y_axis.label bug you pointed out before where it isn't centered when the legend is in Bottom position becomes a bit more urgent to fix now, on one of my test dashboards changing legend mode resulted in:
<img width="725" alt="image" src="https://github.com/perses/perses/assets/9369625/cca9a5dc-423d-403f-9373-c00fff00530a">

sjcobb reviewed (approved):
LGTM! I think the [screenshot](https://github.com/perses/perses/pull/1197#issuecomment-1573023916) I added can be it's own ticket and fixed separately since it's more a problem with how I implemented YAxisLabel [positioning](https://github.com/perses/perses/blob/70bcb3856bc4b53bc3a37abb31d23fcb64ec5aad/ui/components/src/YAxisLabel.tsx#L28) than anything to do with the legend.

Planning to do release today so hopefully this can make it in! 🚀

juliepagano commented:
> LGTM! I think the [screenshot](https://github.com/perses/perses/pull/1197#issuecomment-1573023916) I added can be it's own ticket and fixed separately since it's more a problem with how I implemented YAxisLabel [positioning](https://github.com/perses/perses/blob/70bcb3856bc4b53bc3a37abb31d23fcb64ec5aad/ui/components/src/YAxisLabel.tsx#L28) than anything to do with the legend.
> 
> Planning to do release today so hopefully this can make it in! 🚀

Do you want to hold off on doing a release until early Monday since I think a bunch of the maintainers are out in the afternoon? I think there will be a handful of breaking changes in this one, so it'd be nice to have folks around to give it a close eye.

juliepagano commented:
> The y_axis.label bug you pointed out before where it isn't centered when the legend is in Bottom position becomes a bit more urgent to fix now, on one of my test dashboards changing legend mode resulted in: 
> 
> Also, possibly some z-index adjustments will be needed so that the y_axis label shows on top or gets truncated too in these cases.

Good call! I think we're going to try to give that bug to one of our new teammates next week, since it'll be a nice intro-to-perses task to pick up.
